himbaechel/uarch/gatemate/pack_clocking.cc

```diff
@@ -615,7 +642,186 @@ void GateMatePacker::pack_pll()
+    for (int i = 0; i < 4; i++) {
+        NetInfo *in_net = uarch->clkin[0]->getPort(ctx->idf("CLK%d", i));
+        if (in_net) {
+            if (in_net->driver.cell)
+                clk_iosel[i] = in_net->driver.cell;
+            else
+                use_ser_clk = true;
+        }
+        uarch->clkin[0]->disconnectPort(ctx->idf("CLK%d", i));
+    }
+
+    for (int new_die = 0; new_die < uarch->dies; new_die++) {
+        // Reconnect CLKIN and create appropriate GPIO and IOSEL cells
+        for (int i = 0; i < 4; i++) {
+            if (clk_iosel[i]) {
+                CellInfo *iosel = clk_iosel[i];
+                auto pad_info = uarch->bel_to_pad[iosel->bel];
+                Loc l = uarch->locations[std::make_pair(IdString(pad_info->package_pin), new_die)];
+                CellInfo *iosel_new = ctx->getBoundBelCell(ctx->getBelByLocation(l));
+                if (!iosel_new) {
+                    iosel_new = create_cell_ptr(iosel->type, ctx->idf("%s$die%d", iosel->name.c_str(ctx), new_die));
+                    iosel_new->params = iosel->params;
+                    ctx->bindBel(ctx->getBelByLocation(l), iosel_new, PlaceStrength::STRENGTH_FIXED);
+
+                    CellInfo *gpio = iosel->getPort(id_GPIO_IN)->driver.cell;
+                    CellInfo *gpio_new =
+                            create_cell_ptr(gpio->type, ctx->idf("%s$die%d", gpio->name.c_str(ctx), new_die));
+                    gpio_new->params = gpio->params;
+                    ctx->bindBel(ctx->getBelByLocation({l.x, l.y, 0}), gpio_new, PlaceStrength::STRENGTH_FIXED);
+
+                    // Duplicate input connection
+                    gpio_new->connectPort(id_I, gpio->getPort(id_I));
+                    // Connect IOSEL and CPE_IBUF
+                    gpio_new->connectPorts(id_Y, iosel_new, id_GPIO_IN);
+                }
+                if (iosel_new->getPort(id_IN1))
+                    uarch->clkin[new_die]->connectPort(ctx->idf("CLK%d", i), iosel_new->getPort(id_IN1));
+                else
+                    iosel_new->connectPorts(id_IN1, uarch->clkin[new_die], ctx->idf("CLK%d", i));
+            }
+        }
+        if (use_ser_clk)
+            uarch->clkin[new_die]->connectPort(id_SER_CLK, net_SER_CLK);
+
+        if (new_die != 0) {
+            // Copy configuration from first die to other dies
+            uarch->clkin[new_die]->params = uarch->clkin[0]->params;
+            uarch->glbout[new_die]->params = uarch->glbout[0]->params;
+
+            // Copy PLLs
+            for (int i = 0; i < 4; i++) {
+                Loc fixed_loc = uarch->locations[std::make_pair(ctx->idf("PLL%d", i), 0)];
+                BelId pll_bel = ctx->getBelByLocation(fixed_loc);
+                CellInfo *pll = ctx->getBoundBelCell(pll_bel);
+                if (pll) {
+                    // Create new PLL
+                    CellInfo *pll_new = create_cell_ptr(pll->type, ctx->idf("%s$die%d", pll->name.c_str(ctx), new_die));
```

Comment:
this definitely seems like it could cause some issues if the different PLLs in different dice lock at different times - probably `USR_LOCKED_STDY_RST` (assuming this is the "locked" status output) should be ANDed together between all the dice, and then the output of that AND tree used across all dice. This way all the dice can come out of reset at the same time, if they use the PLL locked output as a reset (as is common).